tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs

```diff
@@ -41,32 +41,48 @@ public void WriteTrace(ArraySegment<Span> trace)
             // We ensure there's no trace (local root span) without a test tag.
             // And ensure all remaining spans have the origin tag.
 
-            HashSet<ulong> removeIds = null;
-            Span[] finalTrace = new Span[trace.Count];
-            int idx = 0;
-            foreach (var span in trace)
+            // **
+            // * Filter entire traces without test or benchmark root span
+            // **
+
+            HashSet<ulong> spanIdsToRemove = null;
```

Comment:
This may be a stupid question, but can't we dramatically simplify this? 

e.g. we know the final span will be the root span. Can we just check that to see if it has a test tag. If yes, add the origin tag to all the spans. If no, then abandon the trace?

Reply:
That was the first approach, but I've been told (CIAPP) to remove the spans from the payload :shrug:

Reply:
>That was the first approach, but I've been told (CIAPP) to remove the spans from the payload

I don't think I understand - if the root span is _not_ a test span, we can immediately bail out, and not send _anything_, so the spans have been removed in that case? 🤔

Reply:
Oh, sorry, I misunderstood, well the algorithm you are describing works when we have a complete trace but the filtering in general is hard when we have partial flushing.

Reply:
>but the filtering in general is hard when we have partial flushing.

Yeah, if you need to support partial flush, that won't work 🙁

Reply:
done in 2a9514d